### PR TITLE
bridge: Add shim channels

### PR DESCRIPTION
**Pull request by petervo**

### CODE REVIEW

File: src/bridge/cockpitrouter.c [hunk 1]
```diff
@@ -55,6 +68,72 @@ enum {
 
 typedef GType (* TypeFunction) (void);
 
+typedef struct
+{
```
Comment (stefwalter) [inline src/bridge/cockpitrouter.c]: Extra wrapping here.

File: src/bridge/cockpitrouter.c [hunk 2]
```diff
@@ -55,6 +68,72 @@ enum {
...
+  CockpitTransport *transport;
+  GHashTable *channels;
+  gchar *id;
+
+  gulong closed_sig;
+  guint timeout;
+
```
Comment (stefwalter) [inline src/bridge/cockpitrouter.c]: Extra line here.

File: src/bridge/cockpitrouter.c [hunk 3]
```diff
@@ -77,12 +156,30 @@ on_channel_closed (CockpitChannel *channel,
...
+  bridge = g_hash_table_lookup (self->bridges_by_channel, id);
+  if (bridge)
+    {
+      g_hash_table_remove (self->bridges_by_channel, id);
+      g_hash_table_remove (bridge->channels, id);
+      if (g_hash_table_size (bridge->channels) == 0)
+        {
+          /*
+           * Close sessions that are no longer in use after N seconds
+           * of them being that way.
```
Comment (stefwalter) [inline src/bridge/cockpitrouter.c]: Out of interest, was this something you imagined would be optional for various bridges? It seems that the superuser privileged bridge would have sideeffects of this cleanup. Are they intentional, and intended to be used in that case?

File: src/bridge/cockpitshim.h [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+#include "cockpitchannel.h"
+
+G_BEGIN_DECLS
+
+#define COCKPIT_TYPE_SHIM         (cockpit_shim_get_type ())
+
+GType              cockpit_shim_get_type   (void) G_GNUC_CONST;
+
+void               cockpit_shim_set_secondary_transport     (CockpitChannel *channel,
+                                                             CockpitTransport *transport);
```
Comment (stefwalter) [inline src/bridge/cockpitshim.h]: This is not defined anywhere.

**Comment by stefwalter** [comment]
I'll make those minor changes and merge.

**Comment by stefwalter** [comment]
Whoops. No, won't merge. Here's a related failure in the tests:

```
Created symlink from /etc/systemd/system/default.target.wants/boot_timer.service to /etc/systemd/system/boot_timer.service.
Unexpected journal message 'send_close_channel: assertion 'self->shim_transport != NULL' failed'
Unexpected journal message 'send_close_channel: assertion 'self->shim_transport != NULL' failed'
Unexpected journal message 'send_close_channel: assertion 'self->shim_transport != NULL' failed'
Unexpected journal message 'send_close_channel: assertion 'self->shim_transport != NULL' failed'
Journal extracted to TestServices-testCreateTimer-10.111.118.6-FAIL.log
Wrote TestServices-testCreateTimer-FAIL.png
Journal extracted to TestServices-testCreateTimer-10.111.118.6-FAIL.log
not ok 71 testCreateTimer (check_services.TestServices) duration: 50s
Traceback (most recent call last):
  File "/build/cockpit/test/common/testlib.py", line 533, in tearDown
    self.check_journal_messages()
  File "/build/cockpit/test/common/testlib.py", line 674, in check_journal_messages
    raise Error(first)
Error: send_close_channel: assertion 'self->shim_transport != NULL' failed
```

File: src/bridge/cockpitrouter.c (hunk 3 above)
Comment (petervo) [inline src/bridge/cockpitrouter.c]: You're right we probably need a way to turn this off for some bridges, can that be a follow up though?

**Comment by petervo** [comment]
Think i fixed that error

File: src/bridge/cockpitrouter.c (hunk 3 above)
Comment (stefwalter) [inline src/bridge/cockpitrouter.c]: Yup. Follow up for sure.

**Pull request closed without merging**